llvm/lib/Target/RISCV/RISCVISelLowering.cpp

```diff
@@ -11582,15 +11621,19 @@ void RISCVTargetLowering::analyzeFormalArgumentsCompute(MachineFunction &MF,
     const bool IsByRef = Arg.hasByRefAttr();
     Type *BaseArgTy = Arg.getType();
     Type *MemArgTy = IsByRef ? Arg.getParamByRefType() : BaseArgTy;
+
+    IntegerType *IntTy = IntegerType::get(Ctx, 32);
+    bool IsSmall = Arg.getType()->getScalarSizeInBits() < 32;
+
     Align Alignment = DL.getValueOrABITypeAlignment(
-        IsByRef ? Arg.getParamAlign() : std::nullopt, MemArgTy);
+        IsByRef ? Arg.getParamAlign() : std::nullopt, IsSmall ? IntTy : MemArgTy);
```

Comment:
IsSmall is always be true while type of arg is charN or shortN, so the alignment is always 4, which leading argOffset will be 4 instead of 8 while the type is char8; is this logic correct?

Reply:
Yes, this is indeed a mistake. Thank you for catching it, I have now modified it as follows: